tests/unit/s2n_client_server_name_extension_test.c

```diff
@@ -0,0 +1,169 @@
+
+        EXPECT_SUCCESS(s2n_stuffer_free(&stuffer));
+        EXPECT_SUCCESS(s2n_connection_free(client_conn));
+        EXPECT_SUCCESS(s2n_connection_free(server_conn));
+    }
+
+    /* recv - extra data ignored */
+    {
+        struct s2n_connection *client_conn;
+        EXPECT_NOT_NULL(client_conn = s2n_connection_new(S2N_CLIENT));
+
+        struct s2n_connection *server_conn;
+        EXPECT_NOT_NULL(server_conn = s2n_connection_new(S2N_SERVER));
+
+        struct s2n_stuffer stuffer;
+        EXPECT_SUCCESS(s2n_stuffer_growable_alloc(&stuffer, 0));
+
+        EXPECT_SUCCESS(s2n_set_server_name(client_conn, test_server_name));
+        EXPECT_SUCCESS(s2n_client_server_name_extension.send(client_conn, &stuffer));
+        EXPECT_SUCCESS(s2n_stuffer_write_uint8(&stuffer, 'a'));
+
+        EXPECT_STRING_EQUAL(server_conn->server_name, "");
+        EXPECT_SUCCESS(s2n_client_server_name_extension.recv(server_conn, &stuffer));
+        EXPECT_STRING_EQUAL(server_conn->server_name, test_server_name);
+
+        EXPECT_SUCCESS(s2n_stuffer_free(&stuffer));
+        EXPECT_SUCCESS(s2n_connection_free(client_conn));
+        EXPECT_SUCCESS(s2n_connection_free(server_conn));
+    }
+
+    /* recv - malformed */
+    {
+        struct s2n_connection *client_conn;
+        EXPECT_NOT_NULL(client_conn = s2n_connection_new(S2N_CLIENT));
+
+        struct s2n_connection *server_conn;
+        EXPECT_NOT_NULL(server_conn = s2n_connection_new(S2N_SERVER));
+
+        struct s2n_stuffer stuffer;
+        EXPECT_SUCCESS(s2n_stuffer_growable_alloc(&stuffer, 0));
+
+        EXPECT_SUCCESS(s2n_set_server_name(client_conn, test_server_name));
+        EXPECT_SUCCESS(s2n_client_server_name_extension.send(client_conn, &stuffer));
+
+        uint8_t extension_size = s2n_stuffer_data_available(&stuffer);
+        uint8_t test_bytes = extension_size - strlen(test_server_name);
+
+        /* Check that inverting any byte in the sizes / name type causes us to skip the extension */
+        for (int i = 0; i < test_bytes; i++ ) {
+            /* Mess something up! */
+            stuffer.blob.data[i] = ~stuffer.blob.data[i];
+
+            EXPECT_STRING_EQUAL(server_conn->server_name, "");
+            EXPECT_SUCCESS(s2n_client_server_name_extension.recv(server_conn, &stuffer));
+            EXPECT_STRING_EQUAL(server_conn->server_name, "");
+
+            EXPECT_SUCCESS(s2n_stuffer_rewrite(&stuffer));
+            EXPECT_SUCCESS(s2n_client_server_name_extension.send(client_conn, &stuffer));
+        }
+
+        /* Check that inverting a byte in the server name itself is fine-- there are
+         * no real rules about the server name! */
+        stuffer.blob.data[test_bytes] = ~stuffer.blob.data[test_bytes];
+
+        EXPECT_STRING_EQUAL(server_conn->server_name, "");
+        EXPECT_SUCCESS(s2n_client_server_name_extension.recv(server_conn, &stuffer));
+        EXPECT_STRING_NOT_EQUAL(server_conn->server_name, "");
```

Comment:
We could be more specific in this check `EXPECT_STRING_EQUAL(server_conn->server_name, stuffer.blob.data);`

Reply:
There isn't any data in the stuffer at this point-- we've read it :)